### ISSUE TITLE
Standard idiomatic JavaScript does not work in libraries

### ISSUE THREAD

**Issue by greggman** [emscripten-core/emscripten#4076]
Example:

```
// library_foo.js
mergeInto(LibraryManager.library, {
  my_js: (function() {
      var callCount = 0;
      return function() {
        console.log("call count:", ++callCount);
      };
  }()),
});

```

```
// foo_test.cpp
#include<stdio.h>
#include<stdlib.h>

extern "C" {
  extern void my_js();
}

int main() {
  my_js();
  return EXIT_SUCCESS;
}
```

the compile and run

```
$ ./emcc --js-library library_foo.js  foo_test.cpp 
$ node a.out.js 
exception thrown: ReferenceError: callCount is not defined,ReferenceError: callCount is not defined
    at _my_js (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:1800:40)
    at _main (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:5850:2)
    at Object.asm._main (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9609:19)
    at Object.callMain (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9699:30)
    at doRun (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9757:60)
    at run (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9771:5)
    at Object.<anonymous> (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9860:1)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)
/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:97
      throw ex;
      ^

ReferenceError: callCount is not defined
    at _my_js (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:1800:40)
    at _main (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:5850:2)
    at Object.asm._main (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9609:19)
    at Object.callMain (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9699:30)
    at doRun (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9757:60)
    at run (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9771:5)
    at Object.<anonymous> (/Users/gregg/Downloads/emsdk_portable/emscripten/1.35.0/a.out.js:9860:1)
    at Module._compile (module.js:434:26)
    at Object.Module._extensions..js (module.js:452:10)
    at Module.load (module.js:355:32)

```

**Comment by juj** [comment]
Even though it might be standard idiomatic, such a feature cannot be supported in `--js-library`, because it would require that the js library files, which are evaluated at compile-time, would have to somehow freeze and export their full JS engine state and migrate that to continue the execution at runtime in the user's browser.

If you want to do something like that, you can use the `--pre-js` or `--post-js` features which import the code as is without transformation and it will get executed as if you had typed it in directly into the file.

The `--js-library` way to export a variable would be to do

	// library_foo.js
	mergeInto(LibraryManager.library, {
	  callCount: 0,
	  my_js__deps: ['callCount'],
	  my_js: (function() {
	      return function() {
	        console.log("call count:", ++_callCount);
	      };
	  }()),
	});

Does either of those work?

**Comment by greggman** [comment]
Those don't seem work as they are now global variables that can be accessed by any code. The point of using a closure is it makes the closed variables 100% private. No possible way for code outside to see the value.

The better question is why are the library files evaluated at compile time? Can't they just be concatenated and not evaluated? Or possibly wrapped and concatenated. Wrapped might be something like

```
// wrapper 
addLibrary(function() {
   $insert_contents_of_a_single_lib_file_here$
});
```

Then at load time `addLibrary` will be called by each library, it can call each function (or keep them in a list and call them later), or something. Basically something like [AMD](https://github.com/amdjs/amdjs-api/blob/master/AMD.md)

**Comment by kripken** [comment]
They are evaluated because we actually run the library files in a JS engine. This lets us run code there that can autogenerate things for the library, which is very convenient in some cases. For example that `mergeInto` is not just boilerplate, but is a call that runs during compilation. Some libraries do more specialized things than a simple call to `mergeInto`.

If you want to run something during startup time, to create a closure variable for example, you can use a `__postset`, for example you could do this:

````
// library_foo.js
mergeInto(LibraryManager.library, {
  my_js__postset: '_my_js();',
  my_js: function() {
    var callCount = 0;
    _my_js = function() {
      console.log("call count:", ++callCount);
    };
  },
});
````

That calls `my_js` during startup, and replaces itself with a method that closes over the `callCount`.

**Comment by greggman** [comment]
Eval isn't really a solution is it? Browsers have wanted to get rid of eval for years. There's even [security headers servers and server to just ban it completely](https://developer.mozilla.org/en-US/docs/Web/Security/CSP/CSP_policy_directives). By requiring eval that's basically requiring the page to allow insecure code. The code emitted by emscripten might be fine but having to allow eval at all means all other content (ads, comment systems, login systems, like buttons, etc...) all expose the page to eval security issues.

Here's another example of ES6 code that doesn't work that nearly everyone who's switching to ES6 is using

```
mergeInto(LibraryManager.library, {
    my_js: () => { console.log("here"); },
});
```

Compiles but doesn't run.

**Comment by kripken** [comment]
`eval`? There is no `eval` in the process here, certainly not in the generated code for the example we are talking about. Maybe there's a misunderstanding somewhere?

**Comment by greggman** [comment]
Sorry, I saw a string with code in it and assumed eval was being used. My bad.

Still the issue isn't specific to that. It's that by executing the code at compile time and then calling `.toSource` or whatever emscripten is doing seems like it's fighting against the platform. See the ES6 example above. Here's another. This is perfectly valid ES6 but doesn't work in emscripten

```
mergeInto(LibraryManager.library, {
    shared: {
        apiKey: "1234",
    },
    my_js() {
         console.log("my_js", this.shared.apiKey);
    },
    my_js2() {
         console.log("my_js2", this.shared.apiKey);
    },
});
```

That `postset` thing seems like a workaround for a decision that's only going to lead to more and more distance between JavaScript and what's allowed to be used with emscripten.

**Comment by juj** [comment]
Again, please use the `--pre-js` or `--post-js` features, or, probably preferred in this case, separate .js files to get fully runtime executed content. Your choice of JavaScript is not limited by what kind of constructs you can pass to `--js-library` files, but you also have the other mechanisms at your disposal.

**Comment by greggman** [comment]
Unfortunately that's not available for me. I'm using  unity3d and they
channel everything though --js-library

Can you provide an example where you code generate? I'm just curious what
special magic is making this incompatible with normal JavaScript practices

**Comment by kripken** [comment]
There isn't any magic going on here: The JS compiler portion is simply a JS program, and it loads multiple files during compilation, it's own modules (utilities, modules, library manager, etc.), as well as each JS library provided. So JS libraries are simply normal JS code that runs during compilation. Any valid JS code for node.js (or whatever you use to run the JS compiler) is valid.

And the Library object is basically a JS object containing `functionName: function`, after running all the JS libraries. That's why if you create a closure there, it's a valid closure while the compiler runs. But not at runtime. (But in the code sample I gave above, you can easily create a closure at runtime too, you just need to add a `postset`, which is just a tiny bit of code that is appended, that makes sure the method is called so the closure is generated.)

So anything that is valid ES6, that can run in node.js - a serious limitation, as it doesn't yet have all ES6 features yet - and that results in `functionName: function` stuff on the Library object, should work. In your example, it looks like it indeed fails because of node.js limitations. Updating node.js might help, if more recent node.js does support that ES6 feature. (Alternatively you can use the SpiderMonkey shell instead of node.js, and that would work there.)

**Comment by greggman** [comment]
The code above runs just fine in node.js, the issue is emscripten
transformed it from valid JavaScript to invalid JavaScript

It started like this

```
{
   my_js() {
      ...
   },
}
```

Which is an object property and valid JavaScript. It gets transformed into

```

my_js() {
    ...
}

```

A stand alone JavaScript call followed by a code block. That's invalid
JavaScript

**Comment by greggman** [comment]
The code above runs just fine in node.js, the issue is emscripten
transformed it from valid JavaScript to invalid JavaScript

It started like this

```
mergeLibrary(..., {
   my_js() {
      ...
   },
});
```
Which is an object property and valid JavaScript. It gets transformed into

```
my_js() {
    ...
}
```

A stand alone JavaScript call followed by a code block. That's invalid JavaScript

**Comment by kripken** [comment]
Oh, you are using a version of node that accepts that syntax? Nice. Ok, then the issue is the second thing I was saying earlier - the way to think about the JS library files is that they are a simple mapping of keys to JS things. They aren't just copied wholesale into the emitted source - we find the necessary things and pull them in one by one, using the keys (so we only bring in what is actually used). The key-value syntax is necessary for that 'linking' process. But, you can still use modern ES6 in an internal object:

````
mergeLibrary(..., {
  $MyJS: {
    someCode() {
      ...
    }
  },

  my_js__deps: ['$MyJS'],
  my_js: function() {
    MyJS.someCode();
  },
});
````

In other words, you can have entire JS objects containing arbitrary ES6, closures, etc. It's just the toplevel that must have the key-value structure. If your toplevel - the C calling interface - just calls into the JS objects and their methods, you can write almost all your code in modern ES6.

**Comment by greggman** [comment]
Okay, thank you

Sorry if I missed it but where is `xxx__deps` and `xxx__postset` documented? Should I try to contribute some or did I just miss it?

**Comment by greggman** [comment]
FYI your example doesn't seem to work

```
mergeLibrary(..., {
  $MyJS: {
    someCode() {
      ...
    }
  },

  my_js__deps: ['$MyJS'],
  my_js: function() {
    MyJS.someCode();
  },
});
```

The ES6 part

```
  $MyJS: {
    someCode() {
      ...
    }
  },
```

is again mis-translated into

```
  var MyJS={someCode:someCode() {
          console.log("someCode");
        }};function _my_js() {
        MyJS.someCode();
      }
```

The some `someCode:` inserted before the `someCode()` is a syntax error

**Comment by kripken** [comment]
Sorry about that, I couldn't test it here - as I said, my version of node.js doesn't support that ES6 feature. It looks like the `toSource()` method on those functions behaves surprisingly. The problem occurs for us in `stringifyWithFunctions`, https://github.com/kripken/emscripten/blob/master/src/utility.js#L355 . What is the typeof of those special functions? Either we need a new type to handle, or for typeof 'function' we need to verify that it starts with `function`, or fix it if not.

In any case, if you want something more direct, you can use a `--pre-js/--post-js` (but you said you can't), or you can try an `EM_ASM` (but it goes through the C preprocessor, so we can make no promises there). A third option, maybe the simplest anyhow. is just to add another script tag to the page. You can write any code in that, and assign to the `Module` object for example, or some other global location. Then the JS library could call into there.

About the docs, yeah, sorry about this, this isn't well documented currently. Looks like `deps` is mentioned in `site/source/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.rst`, but just briefly. I don't see docs for `postset`. Pull request would be great :)
